framework/filters/HostControl.php

```diff
@@ -0,0 +1,166 @@
+ */
+class HostControl extends ActionFilter
+{
+    /**
+     * @var array|\Closure|null list of host names, which are allowed.
+     * Each host can be specified as a wildcard pattern. For example:
+     *
+     * ```php
+     * [
+     *     'example.com',
+     *     '*.example.com',
+     * ]
+     * ```
+     *
+     * This field can be specified as a PHP callback of following signature:
+     *
+     * ```php
+     * function (Action $action) {
+     *     //return array of strings
+     * }
+     * ```
+     *
+     * where `$action` is the current [[Action|action]] object.
+     *
+     * If this field is not set - no host name check will be performed.
+     */
+    public $allowedHosts;
+    /**
+     * @var callable a callback that will be called if the current host does not match [[allowedHosts]].
+     * If not set, [[denyAccess()]] will be called.
+     *
+     * The signature of the callback should be as follows:
+     *
+     * ```php
+     * function (Action $action)
+     * ```
+     *
+     * where `$action` is the current [[Action|action]] object.
+     *
+     * > Note: while implementing your own host deny processing, make sure you avoid usage of the current requested
+     * host name, creation of absolute URL links, caching page parts and so on.
+     */
+    public $denyCallback;
+
+
+    /**
+     * This method is invoked right before an action is to be executed (after all possible filters.)
+     * You may override this method to do last-minute preparation for the action.
+     * @param Action $action the action to be executed.
+     * @return bool whether the action should continue to be executed.
+     */
+    public function beforeAction($action)
+    {
+        $allowedHosts = $this->allowedHosts;
+        if ($allowedHosts instanceof \Closure) {
+            $allowedHosts = call_user_func($allowedHosts, $action);
+        }
+        if ($allowedHosts === null) {
+            return true;
+        }
+
+        if (!is_array($allowedHosts) && !$allowedHosts instanceof \Traversable) {
+            $allowedHosts = (array)$allowedHosts;
+        }
+
+        $currentHost = Yii::$app->getRequest()->getHostName();
+
+        foreach ($allowedHosts as $allowedHost) {
+            if (fnmatch($allowedHost, $currentHost)) {
+                return true;
+            }
+        }
+
+        if ($this->denyCallback !== null) {
+            call_user_func($this->denyCallback, $action);
+        } else {
+            $this->denyAccess($action);
+        }
+
+        return false;
+    }
+
+    /**
+     * Denies the access.
+     * The default implementation will display 404 page right away, terminating the program execution.
+     * @param Action $action the action to be executed.
+     */
+    protected function denyAccess($action)
+    {
+        $response = Yii::$app->getResponse();
+        $errorHandler = Yii::$app->getErrorHandler();
+
+        $exception = new NotFoundHttpException(Yii::t('yii', 'Page not found.'));
```

Comment:
Throwing an exception is not reliable.
See: https://github.com/yiisoft/yii2/pull/13063/files/1a33a69f2f01efe2b0e2e21cb01668a4a6270a52#r89448567